### ISSUE TITLE
multi-GPU training

### ISSUE THREAD

**Issue by shyoulala** [tascj/offload_adam#1]
May I ask if this strategy can seamlessly support multi-GPU training?

**Comment by tascj** [comment]
This implementation includes:

1. BF16 optimizer `offload_adam/adam.py`
2. BF16 optimizer with CPU offloading `offload_adam/offload_adam[_v2].py`

1 could be used like a regular PyTorch optimizer, so multi-GPU training should work fine. I've used it with FSDP without problem.

As for 2 I haven't tried yet. But without state sharding, it will eat a lot of host RAM. But it won't be difficult to implement offloading + ZeRO DP: Every process holds full local gradients and sharded optimizer states, then do reduce-scatter of gradients before opt step and all gather of parameters after opt step.
The integration of 2 is also not that seamless. You need to make a few changes to the training loop, so it's not compatible with those trainer classes out there.

I'm planning to add QAT support to further increase single-GPU training capacity. After that, I'll consider supporting offloading + ZeRO DP.

**Comment by wenmin-wu** [comment]
What's the difference between this one and deepspeed's stage 2 optimiser offload?

**Comment by tascj** [comment]
@wenmin-wu 

Focusing only on the offloading part, the main difference is that this implementation  do weight updates on GPU, while DeepSpeed does it on CPU. This implementation requires transferring more states between host and device, but weight updates are much faster.

I haven't read DeepSpeed's implementation details, but when I tested it last time, the offloading was too slow to be practical. They recently have ZenFlow, which I haven't tested yet.

**Comment by tascj** [comment]
I've added experimental multi-GPU training. It works well on certain workloads and hardware setup. See `examples/multi_gpu/` for usage.

Closing — follow-up issues / PRs welcome.
